{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "618a152e",
   "metadata": {},
   "source": [
    "## Tesseract\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "1313fb00",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "识别的文本内容：\n",
      "由  在何  -人人， 要拥朋值得放 各的轧鸡  就能_\n",
      "\n",
      "发表蒜彩 的 演讲,在友基交潮中叭-  真了生机的\n",
      "_ 友函不是息信,不是舞台展示,也不盘 济介的放言，\n",
      "而是有作值的思故，        _\n",
      "\n",
      "\n"
     ]
    }
   ],
   "source": [
    "import pytesseract\n",
    "from PIL import Image\n",
    "import cv2\n",
    "import numpy as np\n",
    "\n",
    "# 设置 Tesseract 可执行文件的路径（Windows 系统需要）\n",
    "# pytesseract.pytesseract.tesseract_cmd = r'C:\\Program Files\\Tesseract-OCR\\tesseract.exe'\n",
    "\n",
    "def preprocess_image(image_path):\n",
    "    # 读取图像\n",
    "    img = cv2.imread(image_path)\n",
    "    \n",
    "    # 转换为灰度图像\n",
    "    gray = cv2.cvtColor(img, cv2.COLOR_BGR2GRAY)\n",
    "    \n",
    "    # 应用阈值处理以增强对比度\n",
    "    _, thresh = cv2.threshold(gray, 150, 255, cv2.THRESH_BINARY + cv2.THRESH_OTSU)\n",
    "    \n",
    "    return thresh\n",
    "\n",
    "def ocr_from_image(image_path):\n",
    "    try:\n",
    "        # 预处理图像\n",
    "        processed_image = preprocess_image(image_path)\n",
    "        \n",
    "        # 使用 PIL 打开处理后的图像\n",
    "        img = Image.fromarray(processed_image)\n",
    "        \n",
    "        # 使用 Tesseract 进行 OCR 识别\n",
    "        text = pytesseract.image_to_string(img, lang='chi_sim')  # 可根据需要更改语言，如 'chi_sim' 用于中文\n",
    "        \n",
    "        return text\n",
    "    except Exception as e:\n",
    "        return f\"OCR 处理出错: {str(e)}\"\n",
    "\n",
    "# 主函数\n",
    "if __name__ == \"__main__\":\n",
    "    # 替换为你的图像文件路径\n",
    "    image_path = \"/Users/riosky/Pictures/ocr/para.jpeg\"\n",
    "    \n",
    "    # 执行 OCR\n",
    "    result = ocr_from_image(image_path)\n",
    "    \n",
    "    # 打印识别结果\n",
    "    print(\"识别的文本内容：\")\n",
    "    print(result)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "862b0727",
   "metadata": {},
   "source": [
    "## PaddleOCR\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "155ba6ce",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/Users/riosky/miniconda3/envs/rag/lib/python3.11/site-packages/paddle/utils/cpp_extension/extension_utils.py:717: UserWarning: No ccache found. Please be aware that recompiling all source files may be required. You can download and install ccache from: https://github.com/ccache/ccache/blob/master/doc/INSTALL.md\n",
      "  warnings.warn(warning_message)\n",
      "\n",
      "A module that was compiled using NumPy 1.x cannot be run in\n",
      "NumPy 2.2.6 as it may crash. To support both 1.x and 2.x\n",
      "versions of NumPy, modules must be compiled with NumPy 2.0.\n",
      "Some module may need to rebuild instead e.g. with 'pybind11>=2.12'.\n",
      "\n",
      "If you are a user of the module, the easiest solution will be to\n",
      "downgrade to 'numpy<2' or try to upgrade the affected module.\n",
      "We expect that some modules will need time to support NumPy 2.\n",
      "\n",
      "Traceback (most recent call last):  File \"<frozen runpy>\", line 198, in _run_module_as_main\n",
      "  File \"<frozen runpy>\", line 88, in _run_code\n",
      "  File \"/Users/riosky/miniconda3/envs/rag/lib/python3.11/site-packages/ipykernel_launcher.py\", line 18, in <module>\n",
      "    app.launch_new_instance()\n",
      "  File \"/Users/riosky/miniconda3/envs/rag/lib/python3.11/site-packages/traitlets/config/application.py\", line 1075, in launch_instance\n",
      "    app.start()\n",
      "  File \"/Users/riosky/miniconda3/envs/rag/lib/python3.11/site-packages/ipykernel/kernelapp.py\", line 739, in start\n",
      "    self.io_loop.start()\n",
      "  File \"/Users/riosky/miniconda3/envs/rag/lib/python3.11/site-packages/tornado/platform/asyncio.py\", line 211, in start\n",
      "    self.asyncio_loop.run_forever()\n",
      "  File \"/Users/riosky/miniconda3/envs/rag/lib/python3.11/asyncio/base_events.py\", line 608, in run_forever\n",
      "    self._run_once()\n",
      "  File \"/Users/riosky/miniconda3/envs/rag/lib/python3.11/asyncio/base_events.py\", line 1936, in _run_once\n",
      "    handle._run()\n",
      "  File \"/Users/riosky/miniconda3/envs/rag/lib/python3.11/asyncio/events.py\", line 84, in _run\n",
      "    self._context.run(self._callback, *self._args)\n",
      "  File \"/Users/riosky/miniconda3/envs/rag/lib/python3.11/site-packages/ipykernel/kernelbase.py\", line 545, in dispatch_queue\n",
      "    await self.process_one()\n",
      "  File \"/Users/riosky/miniconda3/envs/rag/lib/python3.11/site-packages/ipykernel/kernelbase.py\", line 534, in process_one\n",
      "    await dispatch(*args)\n",
      "  File \"/Users/riosky/miniconda3/envs/rag/lib/python3.11/site-packages/ipykernel/kernelbase.py\", line 437, in dispatch_shell\n",
      "    await result\n",
      "  File \"/Users/riosky/miniconda3/envs/rag/lib/python3.11/site-packages/ipykernel/ipkernel.py\", line 362, in execute_request\n",
      "    await super().execute_request(stream, ident, parent)\n",
      "  File \"/Users/riosky/miniconda3/envs/rag/lib/python3.11/site-packages/ipykernel/kernelbase.py\", line 778, in execute_request\n",
      "    reply_content = await reply_content\n",
      "  File \"/Users/riosky/miniconda3/envs/rag/lib/python3.11/site-packages/ipykernel/ipkernel.py\", line 449, in do_execute\n",
      "    res = shell.run_cell(\n",
      "  File \"/Users/riosky/miniconda3/envs/rag/lib/python3.11/site-packages/ipykernel/zmqshell.py\", line 549, in run_cell\n",
      "    return super().run_cell(*args, **kwargs)\n",
      "  File \"/Users/riosky/miniconda3/envs/rag/lib/python3.11/site-packages/IPython/core/interactiveshell.py\", line 3116, in run_cell\n",
      "    result = self._run_cell(\n",
      "  File \"/Users/riosky/miniconda3/envs/rag/lib/python3.11/site-packages/IPython/core/interactiveshell.py\", line 3171, in _run_cell\n",
      "    result = runner(coro)\n",
      "  File \"/Users/riosky/miniconda3/envs/rag/lib/python3.11/site-packages/IPython/core/async_helpers.py\", line 128, in _pseudo_sync_runner\n",
      "    coro.send(None)\n",
      "  File \"/Users/riosky/miniconda3/envs/rag/lib/python3.11/site-packages/IPython/core/interactiveshell.py\", line 3394, in run_cell_async\n",
      "    has_raised = await self.run_ast_nodes(code_ast.body, cell_name,\n",
      "  File \"/Users/riosky/miniconda3/envs/rag/lib/python3.11/site-packages/IPython/core/interactiveshell.py\", line 3639, in run_ast_nodes\n",
      "    if await self.run_code(code, result, async_=asy):\n",
      "  File \"/Users/riosky/miniconda3/envs/rag/lib/python3.11/site-packages/IPython/core/interactiveshell.py\", line 3699, in run_code\n",
      "    exec(code_obj, self.user_global_ns, self.user_ns)\n",
      "  File \"/var/folders/zn/nh46rr_s0hv7wkhl6f_hbc1c0000gn/T/ipykernel_24874/3868211442.py\", line 1, in <module>\n",
      "    from paddleocr import PaddleOCR\n",
      "  File \"/Users/riosky/miniconda3/envs/rag/lib/python3.11/site-packages/paddleocr-2.7.5-py3.11.egg/paddleocr/__init__.py\", line 14, in <module>\n",
      "    from .paddleocr import *\n",
      "  File \"/Users/riosky/miniconda3/envs/rag/lib/python3.11/site-packages/paddleocr-2.7.5-py3.11.egg/paddleocr/paddleocr.py\", line 21, in <module>\n",
      "    import paddle\n",
      "  File \"/Users/riosky/miniconda3/envs/rag/lib/python3.11/site-packages/paddle/__init__.py\", line 91, in <module>\n",
      "    import paddle.text\n",
      "  File \"/Users/riosky/miniconda3/envs/rag/lib/python3.11/site-packages/paddle/text/__init__.py\", line 15, in <module>\n",
      "    from .datasets import (\n",
      "  File \"/Users/riosky/miniconda3/envs/rag/lib/python3.11/site-packages/paddle/text/datasets/__init__.py\", line 15, in <module>\n",
      "    from .conll05 import Conll05st  # noqa: F401\n",
      "  File \"/Users/riosky/miniconda3/envs/rag/lib/python3.11/site-packages/paddle/text/datasets/conll05.py\", line 27, in <module>\n",
      "    from paddle.dataset.common import _check_exists_and_download\n",
      "  File \"/Users/riosky/miniconda3/envs/rag/lib/python3.11/site-packages/paddle/dataset/__init__.py\", line 18, in <module>\n",
      "    from . import (  # noqa: F401\n",
      "  File \"/Users/riosky/miniconda3/envs/rag/lib/python3.11/site-packages/paddle/dataset/flowers.py\", line 36, in <module>\n",
      "    from paddle.dataset.image import load_image_bytes, simple_transform\n",
      "  File \"/Users/riosky/miniconda3/envs/rag/lib/python3.11/site-packages/paddle/dataset/image.py\", line 39, in <module>\n",
      "    import cv2\n",
      "  File \"/Users/riosky/miniconda3/envs/rag/lib/python3.11/site-packages/opencv_contrib_python-4.6.0.66-py3.11-macosx-11.1-arm64.egg/cv2/__init__.py\", line 181, in <module>\n",
      "    bootstrap()\n",
      "  File \"/Users/riosky/miniconda3/envs/rag/lib/python3.11/site-packages/opencv_contrib_python-4.6.0.66-py3.11-macosx-11.1-arm64.egg/cv2/__init__.py\", line 153, in bootstrap\n",
      "    native_module = importlib.import_module(\"cv2\")\n",
      "  File \"/Users/riosky/miniconda3/envs/rag/lib/python3.11/importlib/__init__.py\", line 126, in import_module\n",
      "    return _bootstrap._gcd_import(name[level:], package, level)\n"
     ]
    },
    {
     "ename": "AttributeError",
     "evalue": "_ARRAY_API not found",
     "output_type": "error",
     "traceback": [
      "\u001b[31m---------------------------------------------------------------------------\u001b[39m",
      "\u001b[31mAttributeError\u001b[39m                            Traceback (most recent call last)",
      "\u001b[31mAttributeError\u001b[39m: _ARRAY_API not found"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "\n",
      "A module that was compiled using NumPy 1.x cannot be run in\n",
      "NumPy 2.2.6 as it may crash. To support both 1.x and 2.x\n",
      "versions of NumPy, modules must be compiled with NumPy 2.0.\n",
      "Some module may need to rebuild instead e.g. with 'pybind11>=2.12'.\n",
      "\n",
      "If you are a user of the module, the easiest solution will be to\n",
      "downgrade to 'numpy<2' or try to upgrade the affected module.\n",
      "We expect that some modules will need time to support NumPy 2.\n",
      "\n",
      "Traceback (most recent call last):  File \"<frozen runpy>\", line 198, in _run_module_as_main\n",
      "  File \"<frozen runpy>\", line 88, in _run_code\n",
      "  File \"/Users/riosky/miniconda3/envs/rag/lib/python3.11/site-packages/ipykernel_launcher.py\", line 18, in <module>\n",
      "    app.launch_new_instance()\n",
      "  File \"/Users/riosky/miniconda3/envs/rag/lib/python3.11/site-packages/traitlets/config/application.py\", line 1075, in launch_instance\n",
      "    app.start()\n",
      "  File \"/Users/riosky/miniconda3/envs/rag/lib/python3.11/site-packages/ipykernel/kernelapp.py\", line 739, in start\n",
      "    self.io_loop.start()\n",
      "  File \"/Users/riosky/miniconda3/envs/rag/lib/python3.11/site-packages/tornado/platform/asyncio.py\", line 211, in start\n",
      "    self.asyncio_loop.run_forever()\n",
      "  File \"/Users/riosky/miniconda3/envs/rag/lib/python3.11/asyncio/base_events.py\", line 608, in run_forever\n",
      "    self._run_once()\n",
      "  File \"/Users/riosky/miniconda3/envs/rag/lib/python3.11/asyncio/base_events.py\", line 1936, in _run_once\n",
      "    handle._run()\n",
      "  File \"/Users/riosky/miniconda3/envs/rag/lib/python3.11/asyncio/events.py\", line 84, in _run\n",
      "    self._context.run(self._callback, *self._args)\n",
      "  File \"/Users/riosky/miniconda3/envs/rag/lib/python3.11/site-packages/ipykernel/kernelbase.py\", line 545, in dispatch_queue\n",
      "    await self.process_one()\n",
      "  File \"/Users/riosky/miniconda3/envs/rag/lib/python3.11/site-packages/ipykernel/kernelbase.py\", line 534, in process_one\n",
      "    await dispatch(*args)\n",
      "  File \"/Users/riosky/miniconda3/envs/rag/lib/python3.11/site-packages/ipykernel/kernelbase.py\", line 437, in dispatch_shell\n",
      "    await result\n",
      "  File \"/Users/riosky/miniconda3/envs/rag/lib/python3.11/site-packages/ipykernel/ipkernel.py\", line 362, in execute_request\n",
      "    await super().execute_request(stream, ident, parent)\n",
      "  File \"/Users/riosky/miniconda3/envs/rag/lib/python3.11/site-packages/ipykernel/kernelbase.py\", line 778, in execute_request\n",
      "    reply_content = await reply_content\n",
      "  File \"/Users/riosky/miniconda3/envs/rag/lib/python3.11/site-packages/ipykernel/ipkernel.py\", line 449, in do_execute\n",
      "    res = shell.run_cell(\n",
      "  File \"/Users/riosky/miniconda3/envs/rag/lib/python3.11/site-packages/ipykernel/zmqshell.py\", line 549, in run_cell\n",
      "    return super().run_cell(*args, **kwargs)\n",
      "  File \"/Users/riosky/miniconda3/envs/rag/lib/python3.11/site-packages/IPython/core/interactiveshell.py\", line 3116, in run_cell\n",
      "    result = self._run_cell(\n",
      "  File \"/Users/riosky/miniconda3/envs/rag/lib/python3.11/site-packages/IPython/core/interactiveshell.py\", line 3171, in _run_cell\n",
      "    result = runner(coro)\n",
      "  File \"/Users/riosky/miniconda3/envs/rag/lib/python3.11/site-packages/IPython/core/async_helpers.py\", line 128, in _pseudo_sync_runner\n",
      "    coro.send(None)\n",
      "  File \"/Users/riosky/miniconda3/envs/rag/lib/python3.11/site-packages/IPython/core/interactiveshell.py\", line 3394, in run_cell_async\n",
      "    has_raised = await self.run_ast_nodes(code_ast.body, cell_name,\n",
      "  File \"/Users/riosky/miniconda3/envs/rag/lib/python3.11/site-packages/IPython/core/interactiveshell.py\", line 3639, in run_ast_nodes\n",
      "    if await self.run_code(code, result, async_=asy):\n",
      "  File \"/Users/riosky/miniconda3/envs/rag/lib/python3.11/site-packages/IPython/core/interactiveshell.py\", line 3699, in run_code\n",
      "    exec(code_obj, self.user_global_ns, self.user_ns)\n",
      "  File \"/var/folders/zn/nh46rr_s0hv7wkhl6f_hbc1c0000gn/T/ipykernel_24874/3868211442.py\", line 1, in <module>\n",
      "    from paddleocr import PaddleOCR\n",
      "  File \"/Users/riosky/miniconda3/envs/rag/lib/python3.11/site-packages/paddleocr-2.7.5-py3.11.egg/paddleocr/__init__.py\", line 14, in <module>\n",
      "    from .paddleocr import *\n",
      "  File \"/Users/riosky/miniconda3/envs/rag/lib/python3.11/site-packages/paddleocr-2.7.5-py3.11.egg/paddleocr/paddleocr.py\", line 25, in <module>\n",
      "    import cv2\n"
     ]
    },
    {
     "ename": "AttributeError",
     "evalue": "_ARRAY_API not found",
     "output_type": "error",
     "traceback": [
      "\u001b[31m---------------------------------------------------------------------------\u001b[39m",
      "\u001b[31mAttributeError\u001b[39m                            Traceback (most recent call last)",
      "\u001b[31mAttributeError\u001b[39m: _ARRAY_API not found"
     ]
    },
    {
     "ename": "ImportError",
     "evalue": "numpy.core.multiarray failed to import",
     "output_type": "error",
     "traceback": [
      "\u001b[31m---------------------------------------------------------------------------\u001b[39m",
      "\u001b[31mImportError\u001b[39m                               Traceback (most recent call last)",
      "\u001b[36mCell\u001b[39m\u001b[36m \u001b[39m\u001b[32mIn[1]\u001b[39m\u001b[32m, line 1\u001b[39m\n\u001b[32m----> \u001b[39m\u001b[32m1\u001b[39m \u001b[38;5;28;01mfrom\u001b[39;00m\u001b[38;5;250m \u001b[39m\u001b[34;01mpaddleocr\u001b[39;00m\u001b[38;5;250m \u001b[39m\u001b[38;5;28;01mimport\u001b[39;00m PaddleOCR\n\u001b[32m      3\u001b[39m \u001b[38;5;66;03m# 初始化 PaddleOCR 实例\u001b[39;00m\n\u001b[32m      4\u001b[39m ocr = PaddleOCR(\n\u001b[32m      5\u001b[39m     use_textline_orientation=\u001b[38;5;28;01mTrue\u001b[39;00m,  \u001b[38;5;66;03m# 启用角度分类（处理旋转文本）\u001b[39;00m\n\u001b[32m      6\u001b[39m     lang=\u001b[33m'\u001b[39m\u001b[33mch\u001b[39m\u001b[33m'\u001b[39m,           \u001b[38;5;66;03m# 设置语言为中文（支持 80+ 种语言，可改为 'en' 等）\u001b[39;00m\n\u001b[32m      7\u001b[39m )\n",
      "\u001b[36mFile \u001b[39m\u001b[32m~/miniconda3/envs/rag/lib/python3.11/site-packages/paddleocr-2.7.5-py3.11.egg/paddleocr/__init__.py:14\u001b[39m\n\u001b[32m      1\u001b[39m \u001b[38;5;66;03m# Copyright (c) 2020 PaddlePaddle Authors. All Rights Reserved.\u001b[39;00m\n\u001b[32m      2\u001b[39m \u001b[38;5;66;03m#\u001b[39;00m\n\u001b[32m      3\u001b[39m \u001b[38;5;66;03m# Licensed under the Apache License, Version 2.0 (the \"License\");\u001b[39;00m\n\u001b[32m   (...)\u001b[39m\u001b[32m     12\u001b[39m \u001b[38;5;66;03m# See the License for the specific language governing permissions and\u001b[39;00m\n\u001b[32m     13\u001b[39m \u001b[38;5;66;03m# limitations under the License.\u001b[39;00m\n\u001b[32m---> \u001b[39m\u001b[32m14\u001b[39m \u001b[38;5;28;01mfrom\u001b[39;00m\u001b[38;5;250m \u001b[39m\u001b[34;01m.\u001b[39;00m\u001b[34;01mpaddleocr\u001b[39;00m\u001b[38;5;250m \u001b[39m\u001b[38;5;28;01mimport\u001b[39;00m *\n\u001b[32m     16\u001b[39m __version__ = paddleocr.VERSION\n\u001b[32m     17\u001b[39m __all__ = [\n\u001b[32m     18\u001b[39m     \u001b[33m'\u001b[39m\u001b[33mPaddleOCR\u001b[39m\u001b[33m'\u001b[39m, \u001b[33m'\u001b[39m\u001b[33mPPStructure\u001b[39m\u001b[33m'\u001b[39m, \u001b[33m'\u001b[39m\u001b[33mdraw_ocr\u001b[39m\u001b[33m'\u001b[39m, \u001b[33m'\u001b[39m\u001b[33mdraw_structure_result\u001b[39m\u001b[33m'\u001b[39m,\n\u001b[32m     19\u001b[39m     \u001b[33m'\u001b[39m\u001b[33msave_structure_res\u001b[39m\u001b[33m'\u001b[39m, \u001b[33m'\u001b[39m\u001b[33mdownload_with_progressbar\u001b[39m\u001b[33m'\u001b[39m, \u001b[33m'\u001b[39m\u001b[33msorted_layout_boxes\u001b[39m\u001b[33m'\u001b[39m,\n\u001b[32m     20\u001b[39m     \u001b[33m'\u001b[39m\u001b[33mconvert_info_docx\u001b[39m\u001b[33m'\u001b[39m, \u001b[33m'\u001b[39m\u001b[33mto_excel\u001b[39m\u001b[33m'\u001b[39m\n\u001b[32m     21\u001b[39m ]\n",
      "\u001b[36mFile \u001b[39m\u001b[32m~/miniconda3/envs/rag/lib/python3.11/site-packages/paddleocr-2.7.5-py3.11.egg/paddleocr/paddleocr.py:25\u001b[39m\n\u001b[32m     21\u001b[39m \u001b[38;5;28;01mimport\u001b[39;00m\u001b[38;5;250m \u001b[39m\u001b[34;01mpaddle\u001b[39;00m\n\u001b[32m     23\u001b[39m sys.path.append(os.path.join(\u001b[34m__dir__\u001b[39m, \u001b[33m'\u001b[39m\u001b[33m'\u001b[39m))\n\u001b[32m---> \u001b[39m\u001b[32m25\u001b[39m \u001b[38;5;28;01mimport\u001b[39;00m\u001b[38;5;250m \u001b[39m\u001b[34;01mcv2\u001b[39;00m\n\u001b[32m     26\u001b[39m \u001b[38;5;28;01mimport\u001b[39;00m\u001b[38;5;250m \u001b[39m\u001b[34;01mlogging\u001b[39;00m\n\u001b[32m     27\u001b[39m \u001b[38;5;28;01mimport\u001b[39;00m\u001b[38;5;250m \u001b[39m\u001b[34;01mnumpy\u001b[39;00m\u001b[38;5;250m \u001b[39m\u001b[38;5;28;01mas\u001b[39;00m\u001b[38;5;250m \u001b[39m\u001b[34;01mnp\u001b[39;00m\n",
      "\u001b[31mImportError\u001b[39m: numpy.core.multiarray failed to import"
     ]
    }
   ],
   "source": [
    "from paddleocr import PaddleOCR\n",
    "\n",
    "# 初始化 PaddleOCR 实例\n",
    "ocr = PaddleOCR(\n",
    "    use_textline_orientation=True,  # 启用角度分类（处理旋转文本）\n",
    "    lang='ch',           # 设置语言为中文（支持 80+ 种语言，可改为 'en' 等）\n",
    ")\n",
    "\n",
    "# 指定图片路径（可以是本地路径或 URL）\n",
    "img_path = '/Users/riosky/Pictures/ocr/para.jpeg'  # 替换为你的图片路径\n",
    "\n",
    "# 执行 OCR 推理\n",
    "result = ocr.predict(img_path)\n",
    "\n",
    "# 输出识别结果\n",
    "for line in result:\n",
    "    box = line[0]           # 文字区域的边界框坐标\n",
    "    text, score = line[1]   # 识别的文本和置信度\n",
    "    print(f\"Text: {text}, Confidence: {score}, Box: {box}\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "9d81e606",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "5f1d26b8",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "rag",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.11.13"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
